controllers/route_controller.go

```diff
@@ -0,0 +1,394 @@
+		}
+
+		err := ctrl.NewControllerManagedBy(mgr).
+			For(routeInfo.gatewayApiType).
+			Watches(&source.Kind{Type: &gateway_api_v1beta1.Gateway{}}, gwEventHandler).
+			Watches(&source.Kind{Type: &corev1.Service{}}, svcEventHandler).
+			Watches(&source.Kind{Type: &mcs_api.ServiceImport{}}, svcImportEventHandler).
+			Complete(&reconciler)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes;httproutes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes/status;httproutes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes/finalizers;httproutes/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Route object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
+func (r *RouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return lattice_runtime.HandleReconcileError(r.reconcile(ctx, req))
+}
+
+func (r *RouteReconciler) reconcile(ctx context.Context, req ctrl.Request) error {
+	route, err := r.getRoute(ctx, req)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if err := r.client.Get(ctx, req.NamespacedName, route.K8sObject()); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if !r.isRouteRelevant(ctx, route) {
+		// not relevant
+		return nil
+	}
+
+	if !route.DeletionTimestamp().IsZero() {
+		r.log.Info("Deleting")
+		r.eventRecorder.Event(route.K8sObject(), corev1.EventTypeNormal,
+			k8s.RouteEventReasonReconcile, "Deleting Reconcile")
+		if err := r.cleanupRouteResources(ctx, route); err != nil {
+			return fmt.Errorf("failed to cleanup GRPCRoute %+v: %w", route, err)
+		}
+		updateRouteListenerStatus(ctx, r.client, route)
+		r.finalizerManager.RemoveFinalizers(ctx, route.K8sObject(), routeTypeToFinalizer[r.routeType])
+
+		// TODO delete metrics
+		return nil
+	} else {
+		r.log.Info("Adding/Updating route")
+		r.eventRecorder.Event(route.K8sObject(), corev1.EventTypeNormal,
+			k8s.RouteEventReasonReconcile, "Adding/Updating Reconcile")
+		err := r.reconcileRouteResource(ctx, route)
+		// TODO add/update metrics
+		return err
+	}
+}
+
+func (r *RouteReconciler) getRoute(ctx context.Context, req ctrl.Request) (core.Route, error) {
+	switch r.routeType {
+	case HTTP:
+		httpRoute := &core.HTTPRoute{}
+		err := r.client.Get(ctx, req.NamespacedName, httpRoute.K8sObject())
+		return httpRoute, err
```

Comment:
I know it's cumbersome to do all if err !=nil return, but lets keep it that way
if err != nil {
  return nil, err
}
return httpRoute,nil

Reply:
Done next commit